Don't skip symlinks during local source upload

dollierp opened the pull request:
# Changes

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

Don't skip symlinks during local source upload

<!-- If this PR fixes a GitHub issue, please mention it like so:

Fixes #<insert issue number here>

-->

/kind bug

# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [x] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)
- [x] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
Content from symlinks is included during local source upload
```

IrvingMg commented on pkg/shp/streamer/tar.go:
```diff
@@ -21,7 +21,7 @@ type Tar struct {
 // skipPath inspect each path and makes sure it skips files the tar helper can't handle.
 func (t *Tar) skipPath(fpath string, stat fs.FileInfo) bool {
-	if !stat.Mode().IsRegular() {
+	if !stat.Mode().IsRegular() && !(stat.Mode()&fs.ModeSymlink != 0) {
```
```suggestion
	if !stat.Mode().IsRegular() && stat.Mode()&fs.ModeSymlink == 0 {
```

nit

IrvingMg commented on pkg/shp/streamer/util.go:
```diff
@@ -28,17 +28,33 @@ func writeFileToTar(tw *tar.Writer, src, fpath string, stat fs.FileInfo) error {
 	}
 
 	header.Name = trimPrefix(src, fpath)
-	if err := tw.WriteHeader(header); err != nil {
-		return err
+
+	// Symlink
+	if stat.Mode()&os.ModeSymlink != 0 {
+		target, err := os.Readlink(fpath)
+		if err != nil {
+			return err
+		}
+
+		header.Linkname = target
```
Should we add validation or a warning if the symlink points outside the source directory?

IrvingMg reviewed (commented):
Thanks for the PR!

Overall, LGTM. I left a couple of comments. It would also be nice to have unit tests for at least basic cases, as the changes are related to the core functionality of the upload command.

dollierp commented:
Hi @IrvingMg,

Thanks for your review.

I updated the PR with new changes trying to address your remarks.

Regards,

IrvingMg commented:
Hi @dollierp, thanks for the changes. 

There’s one more Copilot comment, could you take a look?

dollierp commented:
Hi @IrvingMg,

I updated the PR to make Copilot happy.

Let me know if something else is missing.

Regards.

IrvingMg commented on pkg/shp/streamer/util.go:
```diff
@@ -28,17 +29,73 @@ func writeFileToTar(tw *tar.Writer, src, fpath string, stat fs.FileInfo) error {
+
+		if outside, _ := isSymlinkTargetOutsideOfDir(absSrc, absTarget); outside {
+			relPath, _ := filepath.Rel(src, fpath)
+			fmt.Fprintf(os.Stderr, "WARNING: symlink %q points outside the source directory %q (target: %q)\n", relPath, absSrc, target)
```
Could we send the error through `ioStreams.ErrOut`?

IrvingMg commented on test/symlink/symlink_inside_tree:
```diff
@@ -0,0 +1 @@
+../../test
```
Would it be possible to create and delete the symlink at runtime to avoid committing it?

IrvingMg reviewed (commented):
Thanks again for the changes. I left a couple more comments.

dollierp commented:
Hi @IrvingMg,

I tried to address your latest remarks and updated the PR.

Regards,

IrvingMg commented on pkg/shp/streamer/util.go:
```diff
@@ -21,24 +22,80 @@ func trimPrefix(prefix, fpath string) string {
+	return nil
+}
+
+func isSymlinkTargetOutsideOfDir(dir, target string) (bool, error) {
```
nit: Is it possible to replace this with https://pkg.go.dev/path/filepath#IsLocal?

IrvingMg commented on pkg/shp/streamer/tar_test.go:
```diff
@@ -37,10 +48,21 @@ func Test_Tar(t *testing.T) {
+
+	stdErr := errOut.String()
+
+	t.Run("symlinks", func(_ *testing.T) {
```
Since `*testing.T` is unused, should we drop the `t.Run` wrapper?

IrvingMg commented on pkg/shp/streamer/tar_test.go:
```diff
@@ -3,16 +3,27 @@ package streamer
+	os.Symlink("../test", baseDir+"/test/symlink_inside_tree")
+	os.Symlink("../../foobar/test", baseDir+"/test/symlink_outside_tree")
+
+	defer os.Remove(baseDir + "/test/symlink_inside_tree")
```
Instead of deferring, I think we can use `t.Cleanup`.

IrvingMg commented on pkg/shp/streamer/tar_test.go:
```diff
@@ -3,16 +3,27 @@ package streamer
+	ioStreams, _, _, errOut := genericclioptions.NewTestIOStreams()
+
+	baseDir := "../../.."
+	os.Symlink("../test", baseDir+"/test/symlink_inside_tree")
```
It might be a good idea to check the error here and fail in case of an error. For example:
```go
 if err := os.Symlink("../test", baseDir+"/test/symlink_inside_tree"); err != nil {                                                                                                                                          
      t.Fatalf("setup symlink: %v", err)                                                                                                                                                                                      
 }                                                                                                                                                                                                                           
 t.Cleanup(func() { 
      os.Remove(baseDir + "/test/symlink_inside_tree") 
})
```

IrvingMg reviewed (commented):
Thanks for the update, looks good overall. Just a few final comments from my side.

dollierp commented:
Hi @IrvingMg,

I've tried to incorporate your latest feedback and have updated the pull request.

Regards,

IrvingMg reviewed (commented):
Thanks for the changes! Overall /lgtm

@shipwright-io/approvers PTAL, this PR has been open for a while.

IrvingMg commented:
/lgtm

adambkaplan commented on pkg/shp/streamer/util.go:
```diff
@@ -21,24 +22,76 @@ func trimPrefix(prefix, fpath string) string {
+			}
+		}
+
+		if outside, _ := isSymlinkTargetOutsideOfDir(absSrc, absTarget); outside {
+			relPath, _ := filepath.Rel(t.src, fpath)
+			fmt.Fprintf(t.ioStreams.ErrOut, "WARNING: symlink %q points outside the source directory %q (target: %q)\n", relPath, absSrc, target)
+		}
```
We should fail here, not warn. Reading the contents of a symlink that is outside of a given "directory root" is a very common vulnerability. See [CWE-61](https://cwe.mitre.org/data/definitions/61.html).

adambkaplan commented on pkg/shp/streamer/tar_test.go:
```diff
@@ -3,16 +3,38 @@ package streamer
+		if err := os.Symlink(symlink.target, path); err != nil {
+			t.Fatalf("Couldn't setup test symlink: %v", err)
+		}
+		t.Cleanup(func() { os.Remove(path) })
```
Golangci-lint failing here on an unhandled error from `os.Remove`. I'm fine capturing the error and logging it.


```suggestion
		t.Cleanup(func() { 
		    cleanupErr := os.Remove(path)
		    if cleanupErr != nil {
		        t.Logf("failed to remove path %q: %v", path, err)
		    }
		})
```

adambkaplan commented on pkg/shp/streamer/util.go:
```diff
@@ -21,24 +22,76 @@ func trimPrefix(prefix, fpath string) string {
+	return nil
+}
+
+func isSymlinkTargetOutsideOfDir(dir, target string) (bool, error) {
```
Worth adding a godoc here around CWE-61


```suggestion
// isSymlinkTargetOutsideOfDir checks if the provided link target resolves to an absolute path outside
// of the provided directory. Symlinks to absolute paths outside of a given "root" directory is a common
// attack vector, as these can potentially bypass operating system permission checks that block access to
// sensitive data. See [CWE-61](https://cwe.mitre.org/data/definitions/61.html).
func isSymlinkTargetOutsideOfDir(dir, target string) (bool, error) {
```

adambkaplan reviewed (changes requested):
Code changes generally look good, however I'm going to ask that we turn the existing warning on using symlinks outside the "root" directory into a failure. This is a common source of vulnerabilities. I am actually pretty sure that in Shipwright's predecessor (OpenShift BuildConfig), we had to issue a CVE on this exact issue.

dollierp commented:
Hi @adambkaplan,

Thanks for your review.

I pushed a new commit which should address your concerns.

Regards,

adambkaplan reviewed (approved):
/lgtm

Updates look good

adambkaplan commented on pkg/shp/streamer/util.go:
```diff
@@ -21,24 +22,80 @@ func trimPrefix(prefix, fpath string) string {
+
+		if outside, _ := isSymlinkTargetOutsideOfDir(absSrc, absTarget); outside {
+			relPath, _ := filepath.Rel(t.src, fpath)
+			return fmt.Errorf("symlink %q points outside the source directory %q (target: %q)\n", relPath, absSrc, target)
```
I held back on pointing this out as a nit, but our lint checker is failing because of the trailing newline.


```suggestion
			return fmt.Errorf("symlink %q points outside the source directory %q (target: %q)", relPath, absSrc, target)
```

adambkaplan reviewed (approved):
/approve

Adding "approve" here, now that we are safely handling symlinks that potentially escape the root directory. We'll need at least one more round of review before this is ready for merge.

adambkaplan commented:
/lgtm cancel

Due to failing lint check.

dollierp commented:
Hi @adambkaplan,

I updated the PR to fix the `\n` lint issue.

Regards.

adambkaplan commented:
@dollierp one last item - please [squash your commits](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) using an interactive rebase, and provide a well written [commit message](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages).

Side note for @shipwright-io/contributors - we should add commit squashing and commit message guidance to our [Contributor Guide](https://github.com/shipwright-io/.github/blob/main/CONTRIBUTING.md).

dollierp committed:
Include symbolic links during local source build upload

Until now, the `shp build upload <build-name> /path` command would
silently discard symbolic links when walking through a local source
directory.

This is not backward compatible with the `oc start-build <build-name>
--from-dir=/path` command which supports symbolic links.

This change adds support for symbolic links to Shipwright cli to match
the legacy OpenShift build commands.

To prevent the type of vulnerability described by CWE-61 [1], an additional
security measure is implemented: an error is raised upon detecting a
symbolic link that points to a file outside the provided directory.

[1]: https://cwe.mitre.org/data/definitions/61.html

Signed-off-by: Denis Ollier <dollierp@redhat.com>

dollierp committed:
Include symbolic links during local source build upload

Until now, the `shp build upload <build-name> /path` command would
silently discard symbolic links when walking through a local source
directory.

This is not backward compatible with the `oc start-build <build-name>
--from-dir=/path` command which supports symbolic links.

This change adds support for symbolic links to Shipwright cli to match
the legacy OpenShift build commands.

To prevent the type of vulnerability described by CWE-61 [1], an additional
security measure is implemented: an error is raised upon detecting a
symbolic link that points to a file outside the provided directory.

[1]: https://cwe.mitre.org/data/definitions/61.html

Signed-off-by: Denis Ollier <dollierp@redhat.com>

dollierp commented:
Hi @adambkaplan,

I squashed the commits into a single one and improved its message to match the guidelines.

Regards,

adambkaplan commented:
/lgtm

Failed e2e test appears to be a flake. Re-running (which will gate merge until it passes).
